design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,367 @@
+
+*Enhance Batch retrieval*
+
+At current, the Secrets Provider code looks at the "K8S_SECRET" environment variable in each of the Pods and preforms a batch retrieval against the Conjur API to get the relevant Conjur Secrets. If there is a failure in retrieving one of the secrets, the request fails and no secrets are returned to the request Pod.
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. To avoid this, we will update the code so that if a failure takes place, it will be noted in the logs and a new request will be made, skipping over the Conjur secret that caused the error.
+
+```pseudocode
+GET /secrets{?variable_ids}
+    If error
+        Add erroring secret to list
+        GET /secrets{?variable_ids} with rest of variableIDs
+        Print list of variableIDs that errored
+        return Conjur secrets
+```
+
+
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Elaboration                                                  | Pros                                                         | Cons                                        |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- |
+| 1    | Read all K8s Secrets that are marked with labels and update them with Conjur secrets |                                                              | - Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound |
+| 2    | Get secrets by configuration in Secrets Provider             | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` |                                                              | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we believe the first solution, Read all K8s Secrets that are marked with labels, would be the most K8s native and painless solution for the customer. Also it avoids the repetitive work of having to list all the K8s Secrets each app needs.
```

Comment:
Was this approved secrutiy-wise?

Reply:
Not yet. We are mimicking the current solution just instead of actually defining the secret, we are describing the secret with a label. Do you get a sense that this adds a vulnerability beyond how we handle this in our current offering?